bqskit/passes/processing/treescan.py

```diff
@@ -0,0 +1,190 @@
+"""This module implements the ScanningGateRemovalPass."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+from typing import Callable
+
+from bqskit.compiler.basepass import BasePass
+from bqskit.compiler.passdata import PassData
+from bqskit.ir.circuit import Circuit
+from bqskit.ir.operation import Operation
+from bqskit.ir.opt.cost.functions import HilbertSchmidtResidualsGenerator
+from bqskit.ir.opt.cost.generator import CostFunctionGenerator
+from bqskit.runtime import get_runtime
+from bqskit.utils.typing import is_real_number
+
+_logger = logging.getLogger(__name__)
+
+
+class TreeScanningGateRemovalPass(BasePass):
+    """
+    The ScanningGateRemovalPass class.
+
+    Starting from one side of the circuit, attempt to remove gates one-by-one.
+    """
+
+    def __init__(
+        self,
+        start_from_left: bool = True,
+        success_threshold: float = 1e-8,
+        cost: CostFunctionGenerator = HilbertSchmidtResidualsGenerator(),
+        instantiate_options: dict[str, Any] = {},
+        tree_depth: int = 1,
+        collection_filter: Callable[[Operation], bool] | None = None,
+    ) -> None:
+        """
+        Construct a ScanningGateRemovalPass.
+
+        Args:
+            start_from_left (bool): Determines where the scan starts
+                attempting to remove gates from. If True, scan goes left
+                to right, otherwise right to left. (Default: True)
+
+            success_threshold (float): The distance threshold that
+                determines successful termintation. Measured in cost
+                described by the hilbert schmidt cost function.
+                (Default: 1e-8)
+
+            cost (CostFunction | None): The cost function that determines
+                successful removal of a gate.
+                (Default: HilbertSchmidtResidualsGenerator())
+
+            instantiate_options (dict[str: Any]): Options passed directly
+                to circuit.instantiate when instantiating circuit
+                templates. (Default: {})
+
+            tree_depth (int): The depth of the tree of potential
+                solutions to instantiate. Note that 2^(tree_depth) - 1
+                circuits will be instantiated in parallel.
```

Comment:
What is the default? Can you give a more user-friendly example of what a specific number means? Like what if I set this to three, what does it do there? If this is adequately described in the class docstring, then you can go less verbose here. Also should the default be 1, isn't that just scanning?